narwhals/_pandas_like/utils.py

```diff
@@ -378,75 +378,93 @@ def set_axis(obj: T, index: Any, implementation: str) -> T:
 
 
-def translate_dtype(column: Any) -> DType:
+def __translate_primitive_dtype(dtype: str) -> DType | None:
     from narwhals import dtypes
 
-    dtype = column.dtype
-    if str(dtype) in ("int64", "Int64", "Int64[pyarrow]", "int64[pyarrow]"):
-        return dtypes.Int64()
-    if str(dtype) in ("int32", "Int32", "Int32[pyarrow]", "int32[pyarrow]"):
-        return dtypes.Int32()
-    if str(dtype) in ("int16", "Int16", "Int16[pyarrow]", "int16[pyarrow]"):
-        return dtypes.Int16()
-    if str(dtype) in ("int8", "Int8", "Int8[pyarrow]", "int8[pyarrow]"):
-        return dtypes.Int8()
-    if str(dtype) in ("uint64", "UInt64", "UInt64[pyarrow]", "uint64[pyarrow]"):
-        return dtypes.UInt64()
-    if str(dtype) in ("uint32", "UInt32", "UInt32[pyarrow]", "uint32[pyarrow]"):
-        return dtypes.UInt32()
-    if str(dtype) in ("uint16", "UInt16", "UInt16[pyarrow]", "uint16[pyarrow]"):
-        return dtypes.UInt16()
-    if str(dtype) in ("uint8", "UInt8", "UInt8[pyarrow]", "uint8[pyarrow]"):
-        return dtypes.UInt8()
-    if str(dtype) in (
-        "float64",
-        "Float64",
-        "Float64[pyarrow]",
-        "float64[pyarrow]",
-        "double[pyarrow]",
-    ):
-        return dtypes.Float64()
-    if str(dtype) in (
-        "float32",
-        "Float32",
-        "Float32[pyarrow]",
-        "float32[pyarrow]",
-        "float[pyarrow]",
-    ):
-        return dtypes.Float32()
-    if str(dtype) in (
-        "string",
-        "string[python]",
-        "string[pyarrow]",
-        "large_string[pyarrow]",
-    ):
-        return dtypes.String()
-    if str(dtype) in ("bool", "boolean", "boolean[pyarrow]", "bool[pyarrow]"):
-        return dtypes.Boolean()
-    if str(dtype) in ("category",) or str(dtype).startswith("dictionary<"):
+    dtype_mappers: dict[str, type[dtypes.DType]] = {
+        "int64": dtypes.Int64,
+        "int32": dtypes.Int32,
+        "int16": dtypes.Int16,
+        "int8": dtypes.Int8,
+        "uint64": dtypes.UInt64,
+        "uint32": dtypes.UInt32,
+        "uint16": dtypes.UInt16,
+        "uint8": dtypes.UInt8,
+        "float64": dtypes.Float64,
+        "double": dtypes.Float64,
+        "float32": dtypes.Float32,
+        "float": dtypes.Float32,
+        "string": dtypes.String,
+        "string[python]": dtypes.String,
+        "large_string": dtypes.String,
+        "bool": dtypes.Boolean,
+        "boolean": dtypes.Boolean,
+        "category": dtypes.Categorical,
+        "date32[day]": dtypes.Date,
+    }
+
+    if dtype not in dtype_mappers:
+        return None
+
+    dtype_factory = next(
+        dtype_factory
+        for pandas_dtype, dtype_factory in dtype_mappers.items()
+        if dtype == pandas_dtype
+    )
+    return dtype_factory()
+
+
+def __translate_datetime_dtype(dtype: str) -> DType | None:
+    from narwhals import dtypes
+
+    if dtype in ("category",) or dtype.startswith("dictionary<"):
```

Comment:
I have a couple of questions

- why is category here in `datetime_dtype`?
- why 2 leading underscores? isn't one enough?

Reply:
1. That is a mistake, it should not be there
2. To enforce it is not going to be used outside: https://realpython.com/python-double-underscore/#double-leading-underscore-in-classes-pythons-name-mangling